[WebGPU] Allow bind groups with unused external textures to be used in a shader

tadeuzagallo opened the pull request:
#### 456f3c8e7e526a26a7713ae0b6e2e8ce049a920b
<pre>
[WebGPU] Allow bind groups with unused external textures to be used in a shader
<a href="https://bugs.webkit.org/show_bug.cgi?id=282098">https://bugs.webkit.org/show_bug.cgi?id=282098</a>
<a href="https://rdar.apple.com/138641833">rdar://138641833</a>

Reviewed by Mike Wyrzykowski.

When a bind group member is specified on the API side but not used by the shader
the structs generated by WGSL and the API for the bind group must match. To generate
the correct struct the WGSL compiler uses the information provided by the API, but
so far we used `unsigned&amp;` for all the missing members. The assumption was that all
members would be pointer sized, but that is not correct for texture_external, and
may also not be correct for other types in different platforms. To fix that we
actually use the type information already available to generate the structs on
the WGSL side.

* LayoutTests/fast/webgpu/regression/repro_282098-expected.txt: Added.
* LayoutTests/fast/webgpu/regression/repro_282098.html: Added.
* Source/WebGPU/WGSL/GlobalVariableRewriter.cpp:
(WGSL::RewriteGlobalVariables::insertStructs):
* Source/WebGPU/WGSL/Metal/MetalCodeGenerator.cpp:
(WGSL::Metal::dumpMetalCodeIfNeeded):

Canonical link: <a href="https://commits.webkit.org/286002@main">https://commits.webkit.org/286002@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b7741fb34b571b5535d9f6a5a330c0e9c9eccff2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/74336 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/53765 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/27147 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/78716 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/25574 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/62898 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/1550 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/58430 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/16755 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/77403 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/48594 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/63945 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/38838 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/45580 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/21440 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/23907 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/66985 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/21787 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/80233 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/1653 "Built successfully") | [⏳ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/macOS-Sequoia-Debug-WK2-Tests-EWS "Waiting to run tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/66718 "Found 1 new test failure: inspector/dom/getMediaStats.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/1801 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/63963 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/66001 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16414 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/9942 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/8098 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/1617 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/1646 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/1634 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/1653 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/WebGPU/WGSL/Metal/MetalCodeGenerator.cpp:
```diff
@@ -50,7 +50,7 @@ static StringView metalCodePrologue()
 static void dumpMetalCodeIfNeeded(StringBuilder& stringBuilder)
 {
-    static bool dumpMetalCode = false;
+    static bool dumpMetalCode = true;
```
I think this change should be reverted?

mwyrzykowski reviewed (approved):
Other than logging the shaders by default (which could increase perceived compilation time) - looks good!

tadeuzagallo committed:
[WebGPU] Allow bind groups with unused external textures to be used in a shader
https://bugs.webkit.org/show_bug.cgi?id=282098
rdar://138641833

Reviewed by Mike Wyrzykowski.

When a bind group member is specified on the API side but not used by the shader
the structs generated by WGSL and the API for the bind group must match. To generate
the correct struct the WGSL compiler uses the information provided by the API, but
so far we used `unsigned&` for all the missing members. The assumption was that all
members would be pointer sized, but that is not correct for texture_external, and
may also not be correct for other types in different platforms. To fix that we
actually use the type information already available to generate the structs on
the WGSL side.

* LayoutTests/fast/webgpu/regression/repro_282098-expected.txt: Added.
* LayoutTests/fast/webgpu/regression/repro_282098.html: Added.
* Source/WebGPU/WGSL/GlobalVariableRewriter.cpp:
(WGSL::RewriteGlobalVariables::insertStructs):
* Source/WebGPU/WGSL/Metal/MetalCodeGenerator.cpp:
(WGSL::Metal::dumpMetalCodeIfNeeded):

Canonical link: https://commits.webkit.org/286002@main
